pennylane/ops/op_math/controlled_decompositions.py

```diff
@@ -15,9 +15,107 @@
 This submodule defines functions to decompose controlled operations
 """
 
+import typing
+import numpy as np
+import numpy.linalg as npl
 import pennylane as qml
 from pennylane.operation import Operator
 from pennylane.wires import Wires
+from pennylane import math
+
+
+def _convert_to_su2(U, return_global_phase=False):
+    r"""Convert a 2x2 unitary matrix to :math:`SU(2)`.
+
+    Args:
+        U (array[complex]): A matrix, presumed to be :math:`2 \times 2` and unitary.
+        return_global_phase (bool): If `True`, the return will include
+        the global phase. If `False`, only the :math:`SU(2)` representative
+        is returned.
+
+    Returns:
+        array[complex]: A :math:`2 \times 2` matrix in :math:`SU(2)` that is
+        equivalent to U up to a global phase. If ``return_global_phase=True``,
+        a 2-element tuple is returned, with the first element being the
+        :math:`SU(2)` equivalent and the second, the global phase.
+    """
+    # Compute the determinants
+    dets = math.linalg.det(U)
+
+    exp_angles = math.cast_like(math.angle(dets), 1j) / 2
+    U_SU2 = math.cast_like(U, dets) * math.exp(-1j * exp_angles)
+    return (U_SU2, exp_angles) if return_global_phase else U_SU2
+
+
+def _ensure_real_diagonal_q(q: np.ndarray):
+    """
+    Change the phases of Q so the main diagonal is real, and return the modified Q.
+    """
+    exp_angles = math.angle(math.diag(q))
+    return q * math.exp(-1j * exp_angles).reshape((1, 2))
+
+
+def _matrix_adjoint(matrix: np.ndarray):
+    return math.transpose(math.conj(matrix))
+
+
+def _param_su2(ar: float, ai: float, br: float, bi: float):
+    """
+    Create a matrix in the SU(2) form from complex parameters a, b.
+    The resulting matrix is not guaranteed to be in SU(2), unless |a|^2 + |b|^2 = 1.
+    """
+    return np.array([[complex(ar, ai), complex(-br, bi)], [complex(br, bi), complex(ar, -ai)]])
```

Comment:
Since this is only one line of code, is it really worth extracting into an additional method?

Reply:
doesn't it already break backpropagation support to have the real and imaginary parts separated?